capsules/extra/src/sk68xx.rs

```diff
@@ -0,0 +1,194 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2025.
+
+//! Single-wire LED.
+//!
+//! Tested with the SK68XXMINI on the ESP32-C3-DevKitM-1.
+//!
+//! Datasheet: <https://www.rose-lighting.com/wp-content/uploads/sites/53/2020/05/SK68XX-MINI-HS-REV.04-EN23535RGB-thick.pdf>
+
+use core::cell::Cell;
+
+use kernel::hil::gpio::Pin;
+use kernel::hil::led::Led;
+
+/// The single-wire, tri-color (RGB) LED.
+///
+/// The pulses need to be calibrated based on the clock speed of the chip.
+/// - `T0H`: Number of nops needed for about 0.28 us. This is then scaled for T0L, T1H, and T1L.
+pub struct Sk68xx<'a, P: Pin, const T0H: usize> {
+    pin: &'a P,
+    nop: fn(),
+    red: Cell<bool>,
+    green: Cell<bool>,
+    blue: Cell<bool>,
+}
+
+impl<'a, P: Pin, const T0H: usize> Sk68xx<'a, P, T0H> {
+    pub fn new(pin: &'a P, nop: fn()) -> Self {
+        Self {
+            pin,
+            nop,
+            red: Cell::new(false),
+            green: Cell::new(false),
+            blue: Cell::new(false),
+        }
+    }
+
+    pub fn init(&self) {
+        self.pin.make_output();
+        self.pin.clear();
+        for _ in 0..1000 {
+            (self.nop)();
+        }
+    }
+
+    fn write(&self) {
+        let red = self.red.get();
+        let green = self.green.get();
+        let blue = self.blue.get();
+
+        for i in 0..24 {
+            let high = if i < 8 {
+                // Green
+                green
+            } else if i < 16 {
+                // Red
+                red
+            } else {
+                // Blue
+                blue
+            };
+            if high {
+                // High for 0.74 us
+                self.pin.set();
+                for _ in 0..(T0H * 3) {
+                    (self.nop)();
+                }
+
+                // Low for 0.52 us
+                self.pin.clear();
+                for _ in 0..(T0H * 2) {
+                    (self.nop)();
+                }
+            } else {
+                // High for 0.28 us
+                self.pin.set();
+                for _ in 0..T0H {
+                    (self.nop)();
+                }
+
+                // Low for 0.94 us
+                self.pin.clear();
+                for _ in 0..(T0H * 4) {
+                    (self.nop)();
+                }
+            }
+        }
+
+        // Rest period after to reset the writing and permit a subsequent write
+        // to the LED.
+        for _ in 0..1000 {
```

Comment:
`287 * T0H`? Assumes T0H is no less than 0.28 microseconds.

Should probably be in a function too, because this is used in init() as well. Alternatively, perhaps move this to the beginning of `write()` then you don't need the delay in `init()`?

Reply:
Updated.

Reply:
You fixed `init()`, but you forgot this one in `write()`.